### ISSUE TITLE
Is memory sanitizer incompatible with _FORTIFY_SOURCE=2?

### ISSUE THREAD

**Issue by sitsofe** [google/sanitizers#689]
Running the following program produces an error on Ubuntu 14.04 with a hand compiled clang-3.8 causes sanitizer warnings:

```
// Compile with
// clang -Wall -Wextra -D_GNU_SOURCE=1 -D_FORTIFY_SOURCE=2 -fsanitize=memory -g -O1 fortifytest.c
#include <stdio.h>
#include <string.h>

int main(void) {
        FILE *fd;
        char *fn = "/tmp/notthere";
        char *asfn;

        (void) asprintf(&asfn, "%.*s", (int)strlen(fn), fn);

        printf("Check if asfn is defined via printf (%s)... ", asfn);
        printf("pass\n");
        // Check asfn is defined via fopen
        printf("Check if asfn is defined via fopen... ");
        fd = fopen(asfn, "r");
        printf("pass\n");

        return 0;
}
```

Using less optimisation (`-O0`) or not using `-D_FORTIFY_SOURCE` at all makes the problem go away...

Issue #247 implies people are disabling `_FORTIFY_SOURCE` when using the santizers for a reason. If the sanitizers really are incompatible with it at this stage could a warning be added to the documentation tell people to turn `_FORTIFY_SOURCE` off before compiling?

**Comment by yugr** [comment]
> If the sanitizers really are incompatible with it at this stage
> could a warning be added to the documentation tell people
> to turn _FORTIFY_SOURCE off before compiling?

It's not that they are fundamentally incompatible but rather noone was motivated enough to fix it.

@chefmax

**Comment by yugr** [comment]
Isn't this a dup of #247 btw?

**Comment by sitsofe** [comment]
@yugr:
Nearly. That's asking for the problem to be fixed. I'm just asking for a warning in the docs until such a point that it's fixed because the errors that arise are really weird...

**Comment by chefmax** [comment]
> Isn't this a dup of #247 btw?

It is, perhaps we can close this one?

> It's not that they are fundamentally incompatible but rather noone was motivated enough to fix it.

Yeah, I don't expect sanitizer folks would fix this in near future.

>  I'm just asking for a warning in the docs 

Good enough? (See https://github.com/google/sanitizers/wiki/AddressSanitizer, FAQ section).

**Comment by sitsofe** [comment]
> Good enough? (See https://github.com/google/sanitizers/wiki/AddressSanitizer, FAQ section).

@chefmax: That will do for now - at least there's something.

Perhaps there needs to be a SanitizerCommonFaq/Incompatiblity page? The problem is that AddressSanitizer tends to get all the information and it's only after some digging that you realise the same information applies to the other sanitizers too. For example https://github.com/google/sanitizers/wiki/AddressSanitizer#faq menentions halt_on_error but it applies to MemorySanitizer too (only you don't know that until you read https://github.com/google/sanitizers/wiki/SanitizerCommonFlags ).

**Comment by kcc** [comment]
duping against #247

**Comment by sitsofe** [comment]
I know this issue is closed but I just wanted to note that the [FAQ](https://github.com/google/sanitizers/wiki/AddressSanitizer#faq) now has two warnings about FORTIFY_SOURCE:

> A3: The fortify source compiler feature can mask errors that ASan is able to detect. 

as added by @chefmax and 

> A: Currently ASan (and other sanitizers) doesn't support source fortification, see https://github.com/google/sanitizers/issues/247

as added by @hannob . Could these be merged into one?

**Comment by yugr** [comment]
@chefmax
